lib/src/io.cpp

```diff
@@ -484,45 +359,157 @@ noexcept (false) {
+            if (isexplicit and type == 0 and ofs.explicits.size()) {
+                /*
+                 * Wrap up when we encounter a EFLR of type FILE-HEADER that is
+                 * NOT the first Logical Record. More precisely we expect the
+                 * _first_ LR we encounter to be a FILE-HEADER. We gather up
+                 * this LR and all successive LR's until we encounter another
+                 * FILE-HEADER.
+                 */
+                file.seek( offset );
+                break;
+            }
+            if (isexplicit) ofs.explicits.push_back( offset );
+            /*
+             * Consider doing fdata-indexing on the fly as we are now at the
+             * correct offset to read the OBNAME. That would mean we only need
+             * to traverse the file a single time to index it. Additionally it
+             * would make the caller code from python way nicer.
+             */
+            else            ofs.implicits.push_back( offset );
+        }
+        offset += len;
+    }
+    return ofs;
+}
+
 std::vector< std::pair< std::string, int > >
-findfdata(mio::mmap_source& file,
-          const std::vector< int >& candidates,
-          const std::vector< long long >& tells,
-          const std::vector< int >& residuals)
+findfdata(dl::stream& file, const std::vector< long long >& tells)
 noexcept (false) {
-
-    const auto* ptr = file.data();
-    const auto* end = file.data() + file.size();
     std::vector< std::pair< std::string, int > > xs;
 
-    char name[256] = {};
+    for (auto tell : tells) {
+        char buffer[ 6 ];
+        char lrsh[ DLIS_LRSH_SIZE ];
+
+        int prevsize = 0;
+        int remaining;
+
+        while (true) {
+            /* Read the first byte of the ORIGIN (UVARI) to parse lenght-encoding*/
+            auto nread = file.read(lrsh, tell, DLIS_LRSH_SIZE);
+            if (nread < DLIS_LRSH_SIZE)
+                throw std::runtime_error("findfdata: unable to read LRSH, file may be truncated");
+
+            int len, type;
+            std::uint8_t attrs;
+            dlis_lrsh( lrsh, &len, &attrs, &type );
+
+            remaining = len - DLIS_LRSH_SIZE;
+
+            int n = std::min(remaining, 1);
+            nread = file.read(buffer, n);
+            if (nread < n)
+                throw std::runtime_error("findfdata: unable to read obname, file may be truncated");
+
+            remaining -= n;
+            prevsize += n;
+            if (prevsize == 1) break;
+        }
+
+        const std::uint8_t high = buffer[ 0 ] & 0xC0; // b11000000
 
-    for (auto i : candidates) {
-        const auto tell = tells[i];
-        const auto resi = residuals[i];
-        int offset = resi == 0 ? 8 : 4;
+        int origin_size = 0;
+        switch( high ) {
+            case 0xC0: origin_size = 4; break; // 11
+            case 0x80: origin_size = 2; break; // 10
+            default:   origin_size = 1; break; // 0x
+        }
+
```

Comment:
Variable sized data is a pain! I might agree, not sure yet. The full signature would be either:
`dl::record extract(dl::stream& file, std::int64_t tell)` 
`dl::record extract(dl::stream& file, std::int64_t tell, int n)` 
or just 
`dl::record extract(dl::stream& file, std::int64_t tell, int n)` with a default value for `n` when we want to read the full LR.

bytes_read would be implicitly communicated through `dl::record.data.size()`.

Unsure of the cost of that compared to do it manually as we do now. We will always read too much, but it might not be a big issue.

That could benefit us in other scenarios as well. If LRs are truncated for example, then we can re-read a LR after a failure, with the actually bytes that are there.